src/gradualizer_type.erl

```diff
@@ -0,0 +1,355 @@
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 1996-2018. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+
+-module(gradualizer_type).
+
+-export_type([abstract_type/0]).
```

Comment:
you could also export `abstract_expr`, I think it's not necessary for the definition of `abstract_type` but was also buggy in older Erlang versions, so we can use our own definitions too.
we could also export `abstract_pattern() :: af_pattern()` as it would be useful in `-spec add_types_pats(Pats :: [erl_parse:abstract_expr()],...` (and it is not a subtype of `abstract_expr()`)

Reply:
Why isn't a abstract_pattern() a subtype of abstract_expr()? Is it because they are recursive?